client/src/store/utils/reducerFunctions.js

```diff
@@ -23,6 +23,60 @@ export const addMessageToStore = (state, payload) => {
   });
 };
 
+export const updateMessageReadInStore = (state, senderId, conversationId) => {
+  const conversation = state.filter(conversation=> conversation.id === conversationId)[0];
+  const messages = conversation.messages.map((message) => {
+    if (message.read === false && message.senderId !== senderId) {
+      const messageCopy = {...message}
+      messageCopy.read = true;
+      return messageCopy;
+    } else {
+      return message;
+    }
+  });
+
+  return state.map((conversation)=>{
+    if (conversation.id == conversationId){
+      const newConvo = {...conversation}
+      newConvo.messages = messages;
+      newConvo.messages.forEach(message=>{
+        if (message.id > newConvo.lastReadMessageId && message.read === true && message.senderId !== senderId)
+          {
+            newConvo.lastReadMessageId = message.id
+          }
+      });
+      newConvo.unreadMessagesCount = newConvo.messages.filter(message => message.read === false 
+        && message.senderId === newConvo.otherUser.id).length;
+      return newConvo;
+    } else {
+      return conversation
+    }
+    });
+}
```

Comment:
This logic could be simplified and merged to the `addMessageToStore` reducer function. Note how you're looping two times to find a particular conversation and to update the messages.

Reply:
The two loops are mutually exclusive. One loop is to find a particular conversation. The other one is to loop over `messages` inside of that particular conversation we found from the first loop. Although i can have the messages loop inside of the conversation loop but that just makes code bit uncleaner.
Secondly, I don't see how i can merge it with `addMessageToStore` because here i am updating only a particular field of a message (i.e. read) and also bunch of other stuff (like updating the `unreadCount` and `lastMessageReadId`). Although there is a slight similarity yet the difference b/w these two functions is big enough to keep them seperate

Reply:
Moreover, `addMessageToStore` adds message whenever there is a new message regardless if the user read the message or not. Whereas `updateMessageReadInStore` sends the update when the message is read and changes the `read` status to true Only when it's shown on the screen